### PR TITLE
If fs does not point to a valid td_t, fallback to gs

### PR DESCRIPTION
**Pull request by jhand2**
The OE loader ensures that FS == GS and places the td_t structure at the memory pointed to by FS. Some applications may use/change FS. In the case that FS has been changed, fall back to checking GS.

### CODE REVIEW

File: enclave/core/sgx/td.c [hunk 1]
```diff
@@ -167,6 +167,12 @@ td_t* oe_get_td()
 
     asm("mov %%fs:0, %0" : "=r"(td));
 
+    // The OE loader ensures that FS and GS point to the same place. To make
+    // accesses to td more robust to changes to FS or GS from an application,
+    // check GS if FS does not point to a valid td_t structure.
+    if (!td || td->magic != TD_MAGIC)
```
[the remark this answers is not in the document]
Comment (jhand2) [reply enclave/core/sgx/td.c]: Is there currently a way I check if we're in Windows simulation mode?

The flag to check if we're in simulation mode is stored in the td struct (though it doesn't really need to be) and I don't believe there's currently any mechanism to check if it's Windows simulation mode or Linux simulation mode.

**Comment by anakrish** [comment]
@jhand2 @CodeMonkeyLeet 
There are other possible contracts too:

- "You can change FS, but must restore it before handing control to OE SDK". This is consistent with the Linux ABI expectation that OE SDK has with user code; it does not expect user  code to violate Linux abi (e.g destroying non-volatile registers before returning to OE SDK).

- "You can change FS, but via a special function oe_swap_fs". This is analogous to the [swapgs](https://www.felixcloutier.com/x86/swapgs) instruction that the kernel uses; but it would allow the OE SDK to keep track of FS changes and recover the original FS.

**Comment by anakrish** [comment]
@CodeMonkeyLeet @jhand2 

Given the hardware guarantees provided by SGX, it is possible for the OE SDK to safely recover FS and GS registers without any help from user level code.
I have created a POC PR here: https://github.com/openenclave/openenclave/pull/2637
(not intended for check in). Essentially we can create a "recover-gs-fs" instruction and emulate it to recover FS and GS.

Even though we can do all this; my preference would be that we solve it at the contract level by requiring the user code to restore FS if it modifies it.

**Comment by CodeMonkeyLeet** [comment]
@jhand2 @anakrish
> There are other possible contracts too:
> 
> * "You can change FS, but must restore it before handing control to OE SDK". This is consistent with the Linux ABI expectation that OE SDK has with user code; it does not expect user  code to violate Linux abi (e.g destroying non-volatile registers before returning to OE SDK).
> * "You can change FS, but via a special function oe_swap_fs". This is analogous to the [swapgs](https://www.felixcloutier.com/x86/swapgs) instruction that the kernel uses; but it would allow the OE SDK to keep track of FS changes and recover the original FS.

I would prefer either of these contracts, and reading the Intel SDM, the problem seems to be with exception handling. A recap of SGX behavior:

* On EENTER, host FS/GS is saved to CR_SAVE_FS/GS and enclave FS/GS is set using TCS.OFSBASGX/OGSBASGX.
  * The TCS values are set during enclave creation per thread and should be immutable.
* On EEXIT, host FS/GS is restored from CR_SAVE_FS/GS
* On AEX, enclave FS/GS is saved to CSSA, host FS/GS is restored from CR_SAVE_FS/GS
  * SGX supports WRFSBASE and WRGSBASE from inside enclaves on 64-bit
* On ERESUME, host FS/GS is saved to CR_SAVE_FS/GS and enclave FS/GS is restored from CSSA.

So far so good. The problem comes in OE SDK exception handling feature. To recap, it works by:

1. Host registers a signal handler, which calls the enclave .
2. On AEX, host signal handler checks if it originals from AEP, if so, it EENTERs to call `oe_virtual_exception_dispatcher()`.
3. `oe_virtual_exception_dispatcher()` only checks for exception context in the current thread, and sets the resume point in the enclave to `oe_real_exception_dispatcher()` if found. It then marks the exception as handled and returns via EEXIT to host.
4. Host signal handler returns to AEP which simply ERESUMEs to the modified `oe_real_exception_dispatcher()`.

The problem is at that point, `oe_real_exception_dispatcher()` still assumes that it has the OE SDK FS to manipulate its TD, but ERESUME would have restored it to the last WRFSBASE set value. Modifying `oe_real_exception_dispatcher()` to find TD from TCS instead of FS would solve this problem. 

Related to #2626, except that instead of handling EENTER and EEXIT, it feels like we only need to address the narrow case of exception handling, and implementing either of the two contracts Anand suggested.
* In the first case, there's a bit more work since anytime an app calls out, they need to remember to save and restore their own FS, including in any enclave exception handlers they've registered in OE (since while ERESUME will restore their state for them before their handler runs, if it makes an OCALL, it's still not going to get the expected FS back on return from that call).
* In the second case, `oe_swap_fs()` would allow OE (and EDL generated stubs) to correctly restore FS on an app's behalf on return from OCALL, and also fix up its own TD lookup in `oe_real_exception_dispatcher()`. The only thing I'm not fond of is actually having OE SDK be responsible for the WRGSBASE invocation, because that requires CR4.FSGSBASE to be set outside of the enclave by a driver. Maybe something like `oe_update_virtual_fs()` might be a better contract.

**Comment by anakrish** [comment]
@CodeMonkeyLeet Nice analysis of the exception dispatcher.

This discussion reveals that we need to clearly layout what an `oe_vectored_exception_handler_t` is allowed to do. Posix/Linux call this [Signal Safety](http://www.man7.org/linux/man-pages/man7/signal-safety.7.html).
Note: Historically signal handlers we allowed to do very limited things; it is only recently (2013) that they could even start using `pthread_self`:
>        pthread_kill(3)        Added in POSIX.1-2013
>        pthread_self(3)        Added in POSIX.1-2013
>        pthread_sigmask(3)     Added in POSIX.1-2013

So, it is not expected that signal handlers can safely use FS unless the library they are written to supports POSIX.1-2013.

In any case, we need to explicitly layout what `oe_vectored_exception_handler_t` signal handlers can do/not do.

Do you happen to know why we invoke user-written signal handlers in the second-pass exception handler https://github.com/openenclave/openenclave/blob/5a5a2bd0b0e4e63243b1f19979b36c050f252153/enclave/core/sgx/exception.c#L273

rather than the first-pass exception handler, as we do for emulating cpuid:
https://github.com/openenclave/openenclave/blob/5a5a2bd0b0e4e63243b1f19979b36c050f252153/enclave/core/sgx/exception.c#L360

?

**Comment by CodeMonkeyLeet** [comment]
@anakrish 
> Do you happen to know why we invoke user-written signal handlers in the second-pass exception handler rather than the first-pass exception handler, as we do for emulating cpuid?

No, I don't recall the reasons for this design; part of it is certainly that we based it on the Linux Intel SGX SDK approach, which also uses a two-pass method. This was also back when permitted nested OCALLS, so allowing a user-authored exception handler to run in the context of a nested call with an outstanding AEX probably creates all kinds of issues. I also don't know if signals remain queued while handling an AEX, so there may be a desire to minimize the window prior to ERESUME.

**Comment by anakrish** [comment]
@jhand2 Do you still need this PR?

**Comment by jhand2** [comment]
> @jhand2 Do you still need this PR?

I'll probably end up solving the problem a different way, so no.

**Pull request closed without merging**